File: go/models/models_test.go
// SPDX-License-Identifier: MIT

package models_test

import (
	"testing"
	"time"

	"github.com/thegeeknetwork/aether-protocol-go/models"
)

// ── NodeCapabilities ──────────────────────────────────────────────────────────

func TestNodeCapabilities_SingleBits(t *testing.T) {
	caps := []struct {
		name string
		cap  models.NodeCapabilities
		want models.NodeCapabilities
	}{
		{"CapabilityBLE", models.CapabilityBLE, 1},
		{"CapabilityWifiDirect", models.CapabilityWifiDirect, 2},
		{"CapabilityGateway", models.CapabilityGateway, 4},
		{"CapabilityRelay", models.CapabilityRelay, 8},
		{"CapabilitySos", models.CapabilitySos, 16},
		{"CapabilityStreaming", models.CapabilityStreaming, 32},
		{"CapabilityVoice", models.CapabilityVoice, 64},
		{"CapabilityDtnCarrier", models.CapabilityDtnCarrier, 128},
		{"CapabilityNearLink", models.CapabilityNearLink, 256},
		{"CapabilityVideo", models.CapabilityVideo, 512},
	}
	for _, tc := range caps {
		if tc.cap != tc.want {
			t.Errorf("%s: got %d, want %d", tc.name, tc.cap, tc.want)
		}
	}
}

func TestNodeCapabilities_BitwiseCombination(t *testing.T) {
	combined := models.CapabilityBLE | models.CapabilityRelay | models.CapabilitySos
	if combined&models.CapabilityBLE == 0 {
		t.Error("BLE bit should be set")
	}
	if combined&models.CapabilityRelay == 0 {
		t.Error("Relay bit should be set")
	}
	if combined&models.CapabilitySos == 0 {
		t.Error("SOS bit should be set")
	}
	if combined&models.CapabilityGateway != 0 {
		t.Error("Gateway bit should NOT be set")
	}
}

func TestNodeCapabilities_AllSet(t *testing.T) {
	all := models.CapabilityBLE | models.CapabilityWifiDirect | models.CapabilityGateway |
		models.CapabilityRelay | models.CapabilitySos | models.CapabilityStreaming |
		models.CapabilityVoice | models.CapabilityDtnCarrier | models.CapabilityNearLink |
		models.CapabilityVideo
	if all != 1023 {
		t.Errorf("all capabilities OR'd: got %d, want 1023", all)
	}
}

func TestNodeCapabilities_UnderlyingType(t *testing.T) {
	// Must be a uint16, so it can hold all 10 capability bits (≤ 65535)
	var cap models.NodeCapabilities = 1023
	if cap != 1023 {
		t.Error("NodeCapabilities should hold 1023 without overflow")
	}
}

// ── AetherNode ────────────────────────────────────────────────────────────────

func TestAetherNode_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	node := models.AetherNode{
		UHID:             "test-uhid-123",
		IdentityKey:      []byte("32-byte-ed25519-identity-key-xxx"),
		Capabilities:     models.CapabilityBLE | models.CapabilityRelay,
		IsLocal:          true,
		LastSeen:         now,
		ReliabilityScore: 85,
	}
	if node.UHID != "test-uhid-123" {
		t.Errorf("UHID: got %q", node.UHID)
	}
	if len(node.IdentityKey) != 32 {
		t.Errorf("IdentityKey length: got %d", len(node.IdentityKey))
	}
	if node.Capabilities&models.CapabilityBLE == 0 {
		t.Error("BLE capability should be set")
	}
	if !node.IsLocal {
		t.Error("IsLocal should be true")
	}
	if node.ReliabilityScore != 85 {
		t.Errorf("ReliabilityScore: got %d", node.ReliabilityScore)
	}
}

func TestAetherNode_ZeroValue(t *testing.T) {
	var node models.AetherNode
	if node.UHID != "" {
		t.Error("zero-value UHID should be empty string")
	}
	if node.IsLocal {
		t.Error("zero-value IsLocal should be false")
	}
	if node.Capabilities != 0 {
		t.Error("zero-value Capabilities should be 0")
	}
}

// ── PeerInfo ──────────────────────────────────────────────────────────────────

func TestPeerInfo_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	peer := models.PeerInfo{
		UHID:             "peer-456",
		Addresses:        []string{"192.168.1.1:8080", "ble://aa:bb:cc"},
		Capabilities:     models.CapabilityWifiDirect,
		LastSeen:         now,
		HopCount:         3,
		ReliabilityScore: 70,
	}
	if peer.UHID != "peer-456" {
		t.Errorf("UHID: got %q", peer.UHID)
	}
	if len(peer.Addresses) != 2 {
		t.Errorf("Addresses count: got %d", len(peer.Addresses))
	}
	if peer.HopCount != 3 {
		t.Errorf("HopCount: got %d", peer.HopCount)
	}
}

// ── RouteEntry.IsStale ────────────────────────────────────────────────────────

func TestRouteEntry_IsStale_Expired(t *testing.T) {
	re := models.RouteEntry{
		DestinationUhid: "dest",
		NextHop:         "hop",
		HopCount:        2,
		ExpiresAt:       time.Now().Add(-1 * time.Second), // already expired
		QualityScore:    90,
		SourceUhid:      "src",
	}
	if !re.IsStale() {
		t.Error("route expired 1s ago should be stale")
	}
}

func TestRouteEntry_IsStale_NotYetExpired(t *testing.T) {
	re := models.RouteEntry{
		DestinationUhid: "dest",
		NextHop:         "hop",
		HopCount:        1,
		ExpiresAt:       time.Now().Add(60 * time.Second),
		QualityScore:    100,
		SourceUhid:      "src",
	}
	if re.IsStale() {
		t.Error("route expiring in 60s should not be stale")
	}
}

func TestRouteEntry_IsStale_ExactlyNow(t *testing.T) {
	// A route expiring exactly at time.Now() — behaviour: time.Now().After(expiresAt) ≈ false
	// (this test is inherently racy; just verify no panic)
	re := models.RouteEntry{ExpiresAt: time.Now()}
	_ = re.IsStale() // must not panic
}

func TestRouteEntry_IsStale_FarFuture(t *testing.T) {
	re := models.RouteEntry{ExpiresAt: time.Now().Add(24 * time.Hour)}
	if re.IsStale() {
		t.Error("route expiring tomorrow should not be stale")
	}
}

// ── DtnBundle.IsExpired ───────────────────────────────────────────────────────

func TestDtnBundle_IsExpired_True(t *testing.T) {
	b := models.DtnBundle{
		ID:               "bundle-001",
		SenderUhid:       "alice",
		RecipientUhid:    "bob",
		EncryptedPayload: []byte("data"),
		ExpiresAt:        time.Now().Add(-5 * time.Minute),
	}
	if !b.IsExpired() {
		t.Error("bundle expired 5m ago should be expired")
	}
}

func TestDtnBundle_IsExpired_False(t *testing.T) {
	b := models.DtnBundle{
		ID:               "bundle-002",
		SenderUhid:       "alice",
		RecipientUhid:    "bob",
		EncryptedPayload: []byte("data"),
		ExpiresAt:        time.Now().Add(30 * time.Minute),
	}
	if b.IsExpired() {
		t.Error("bundle expiring in 30m should not be expired")
	}
}

func TestDtnBundle_IsExpired_AllFieldsPreserved(t *testing.T) {
	now := time.Now().UTC()
	b := models.DtnBundle{
		ID:                   "bndl-xyz",
		SenderUhid:           "s",
		RecipientUhid:        "r",
		EncryptedPayload:     []byte{0xAB, 0xCD},
		Priority:             models.DtnPriorityHigh,
		Status:               models.DtnStatusInCustody,
		CopyCount:            3,
		MaxCopies:            5,
		SenderGeohash:        "abcdef",
		RecipientLastGeohash: "ghijkl",
		HopCount:             2,
		CreatedAt:            now,
		ExpiresAt:            now.Add(time.Hour),
	}
	if b.Priority != models.DtnPriorityHigh {
		t.Error("Priority not preserved")
	}
	if b.Status != models.DtnStatusInCustody {
		t.Error("Status not preserved")
	}
	if b.CopyCount != 3 || b.MaxCopies != 5 {
		t.Error("CopyCount/MaxCopies not preserved")
	}
}

// ── DtnPriority constants ─────────────────────────────────────────────────────

func TestDtnPriority_Values(t *testing.T) {
	if models.DtnPriorityLow != 0 {
		t.Errorf("DtnPriorityLow: got %d, want 0", models.DtnPriorityLow)
	}
	if models.DtnPriorityNormal != 1 {
		t.Errorf("DtnPriorityNormal: got %d, want 1", models.DtnPriorityNormal)
	}
	if models.DtnPriorityHigh != 2 {
		t.Errorf("DtnPriorityHigh: got %d, want 2", models.DtnPriorityHigh)
	}
	if models.DtnPrioritySos != 3 {
		t.Errorf("DtnPrioritySos: got %d, want 3", models.DtnPrioritySos)
	}
}

func TestDtnPriority_Ordering(t *testing.T) {
	if !(models.DtnPriorityLow < models.DtnPriorityNormal &&
		models.DtnPriorityNormal < models.DtnPriorityHigh &&
		models.DtnPriorityHigh < models.DtnPrioritySos) {
		t.Error("DtnPriority constants should be in increasing order")
	}
}

// ── DtnStatus constants ───────────────────────────────────────────────────────

func TestDtnStatus_Values(t *testing.T) {
	cases := []struct {
		name string
		got  models.DtnStatus
		want models.DtnStatus
	}{
		{"Pending", models.DtnStatusPending, 0},
		{"InCustody", models.DtnStatusInCustody, 1},
		{"Delivered", models.DtnStatusDelivered, 2},
		{"Expired", models.DtnStatusExpired, 3},
		{"Failed", models.DtnStatusFailed, 4},
	}
	for _, tc := range cases {
		if tc.got != tc.want {
			t.Errorf("DtnStatus%s: got %d, want %d", tc.name, tc.got, tc.want)
		}
	}
}

// ── PresenceStatus constants ──────────────────────────────────────────────────

func TestPresenceStatus_Values(t *testing.T) {
	if models.PresenceOnline != 0 {
		t.Errorf("PresenceOnline: got %d, want 0", models.PresenceOnline)
	}
	if models.PresenceBusy != 1 {
		t.Errorf("PresenceBusy: got %d, want 1", models.PresenceBusy)
	}
	if models.PresenceAway != 2 {
		t.Errorf("PresenceAway: got %d, want 2", models.PresenceAway)
	}
	if models.PresenceOffline != 3 {
		t.Errorf("PresenceOffline: got %d, want 3", models.PresenceOffline)
	}
}

// ── SosAlert ─────────────────────────────────────────────────────────────────

func TestSosAlert_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	alert := models.SosAlert{
		ID:            "sos-001",
		SenderUhid:    "alice",
		BroadcastType: "sos",
		Message:       "Need help!",
		Latitude:      -26.2041,
		Longitude:     28.0473,
		Geohash:       "ke7fq5",
		Timestamp:     now,
		ReceivedAt:    now,
	}
	if alert.ID != "sos-001" {
		t.Errorf("ID: got %q", alert.ID)
	}
	if alert.SenderUhid != "alice" {
		t.Errorf("SenderUhid: got %q", alert.SenderUhid)
	}
	if alert.Latitude != -26.2041 {
		t.Errorf("Latitude: got %f", alert.Latitude)
	}
	if alert.Longitude != 28.0473 {
		t.Errorf("Longitude: got %f", alert.Longitude)
	}
	if alert.Message != "Need help!" {
		t.Errorf("Message: got %q", alert.Message)
	}
}

// ── CustodyRecord ─────────────────────────────────────────────────────────────

func TestCustodyRecord_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	rec := models.CustodyRecord{
		ID:            "cust-001",
		BundleID:      "bundle-abc",
		FromUhid:      "node-a",
		ToUhid:        "node-b",
		Accepted:      true,
		TransferredAt: now,
	}
	if rec.ID != "cust-001" {
		t.Errorf("ID: got %q", rec.ID)
	}
	if rec.BundleID != "bundle-abc" {
		t.Errorf("BundleID: got %q", rec.BundleID)
	}
	if !rec.Accepted {
		t.Error("Accepted should be true")
	}
	if !rec.TransferredAt.Equal(now) {
		t.Error("TransferredAt not preserved")
	}
}

func TestCustodyRecord_Rejected(t *testing.T) {
	rec := models.CustodyRecord{
		ID:       "cust-002",
		BundleID: "bundle-xyz",
		FromUhid: "node-c",
		ToUhid:   "node-d",
		Accepted: false,
	}
	if rec.Accepted {
		t.Error("Accepted should be false when custody is rejected")
	}
}

// ── DtnDeliveryReceipt ────────────────────────────────────────────────────────

func TestDtnDeliveryReceipt_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	receipt := models.DtnDeliveryReceipt{
		BundleID:              "bundle-001",
		RecipientUhid:         "bob",
		TotalHops:             5,
		TotalCustodyTransfers: 3,
		DeliveredAt:           now,
	}
	if receipt.BundleID != "bundle-001" {
		t.Errorf("BundleID: got %q", receipt.BundleID)
	}
	if receipt.RecipientUhid != "bob" {
		t.Errorf("RecipientUhid: got %q", receipt.RecipientUhid)
	}
	if receipt.TotalHops != 5 {
		t.Errorf("TotalHops: got %d", receipt.TotalHops)
	}
	if receipt.TotalCustodyTransfers != 3 {
		t.Errorf("TotalCustodyTransfers: got %d", receipt.TotalCustodyTransfers)
	}
}

// ── PresenceBeacon ────────────────────────────────────────────────────────────

func TestPresenceBeacon_FieldAssignment(t *testing.T) {
	now := time.Now().UTC()
	beacon := models.PresenceBeacon{
		UHID:          "node-xyz",
		Status:        models.PresenceOnline,
		StatusMessage: "Available",
		Timestamp:     now,
		Geohash:       "ke7fq5",
	}
	if beacon.UHID != "node-xyz" {
		t.Errorf("UHID: got %q", beacon.UHID)
	}
	if beacon.Status != models.PresenceOnline {
		t.Errorf("Status: got %d", beacon.Status)
	}
	if beacon.StatusMessage != "Available" {
		t.Errorf("StatusMessage: got %q", beacon.StatusMessage)
	}
}

File: kotlin/src/test/kotlin/aether/models/ModelsTest.kt
// SPDX-License-Identifier: MIT

package aether.models

import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.UUID
import kotlin.test.*

// ── NodeCapabilities.toBitfield / fromBitfield ────────────────────────────────

class NodeCapabilitiesTest {

    @Test fun `default capabilities all false, bitfield is 0`() {
        assertEquals(0, NodeCapabilities().toBitfield())
    }

    @Test fun `ble flag sets bit 1`() {
        assertEquals(1, NodeCapabilities(ble = true).toBitfield())
    }

    @Test fun `wifiDirect flag sets bit 2`() {
        assertEquals(2, NodeCapabilities(wifiDirect = true).toBitfield())
    }

    @Test fun `gateway flag sets bit 4`() {
        assertEquals(4, NodeCapabilities(gateway = true).toBitfield())
    }

    @Test fun `relay flag sets bit 8`() {
        assertEquals(8, NodeCapabilities(relay = true).toBitfield())
    }

    @Test fun `sos flag sets bit 16`() {
        assertEquals(16, NodeCapabilities(sos = true).toBitfield())
    }

    @Test fun `streaming flag sets bit 32`() {
        assertEquals(32, NodeCapabilities(streaming = true).toBitfield())
    }

    @Test fun `voice flag sets bit 64`() {
        assertEquals(64, NodeCapabilities(voice = true).toBitfield())
    }

    @Test fun `dtnCarrier flag sets bit 128`() {
        assertEquals(128, NodeCapabilities(dtnCarrier = true).toBitfield())
    }

    @Test fun `nearLink flag sets bit 256`() {
        assertEquals(256, NodeCapabilities(nearLink = true).toBitfield())
    }

    @Test fun `video flag sets bit 512`() {
        assertEquals(512, NodeCapabilities(video = true).toBitfield())
    }

    @Test fun `all flags set produces 1023`() {
        val all = NodeCapabilities(
            ble = true, wifiDirect = true, gateway = true, relay = true,
            sos = true, streaming = true, voice = true, dtnCarrier = true,
            nearLink = true, video = true
        )
        assertEquals(1023, all.toBitfield())
    }

    @Test fun `fromBitfield 0 produces all-false capabilities`() {
        val caps = NodeCapabilities.fromBitfield(0)
        assertFalse(caps.ble)
        assertFalse(caps.wifiDirect)
        assertFalse(caps.relay)
        assertFalse(caps.video)
    }

    @Test fun `fromBitfield 1023 produces all-true capabilities`() {
        val caps = NodeCapabilities.fromBitfield(1023)
        assertTrue(caps.ble)
        assertTrue(caps.wifiDirect)
        assertTrue(caps.gateway)
        assertTrue(caps.relay)
        assertTrue(caps.sos)
        assertTrue(caps.streaming)
        assertTrue(caps.voice)
        assertTrue(caps.dtnCarrier)
        assertTrue(caps.nearLink)
        assertTrue(caps.video)
    }

    @Test fun `toBitfield and fromBitfield are inverse`() {
        val original = NodeCapabilities(ble = true, relay = true, voice = true)
        val bitfield = original.toBitfield()
        val restored = NodeCapabilities.fromBitfield(bitfield)
        assertEquals(original, restored)
    }

    @Test fun `fromBitfield round-trips for single flags`() {
        for (bit in 0..9) {
            val bits = 1 shl bit
            val caps = NodeCapabilities.fromBitfield(bits)
            assertEquals(bits, caps.toBitfield(), "round-trip failed for bit $bit (value $bits)")
        }
    }
}

// ── RouteEntry.isExpired ──────────────────────────────────────────────────────

class RouteEntryTest {

    @Test fun `isExpired returns true when expiry is in the past`() {
        val entry = RouteEntry(
            destinationUhid = "dest",
            nextHopUhid = "hop",
            hopCount = 2,
            expiresAt = Instant.now().minusSeconds(10)
        )
        assertTrue(entry.isExpired())
    }

    @Test fun `isExpired returns false when expiry is in the future`() {
        val entry = RouteEntry(
            destinationUhid = "dest",
            nextHopUhid = "hop",
            hopCount = 1,
            expiresAt = Instant.now().plusSeconds(60)
        )
        assertFalse(entry.isExpired())
    }

    @Test fun `isExpired false for far-future expiry`() {
        val entry = RouteEntry(
            destinationUhid = "dest",
            nextHopUhid = "hop",
            hopCount = 1,
            expiresAt = Instant.now().plusSeconds(86400)
        )
        assertFalse(entry.isExpired())
    }

    @Test fun `default expiry is in the future`() {
        val entry = RouteEntry(destinationUhid = "d", nextHopUhid = "h", hopCount = 1)
        assertFalse(entry.isExpired(), "default route should not be immediately expired")
    }
}

// ── DtnBundle.isExpired ───────────────────────────────────────────────────────

class DtnBundleTest {

    @Test fun `isExpired returns true for past expiry`() {
        val b = DtnBundle(
            senderUhid = "alice",
            recipientUhid = "bob",
            encryptedPayload = byteArrayOf(1, 2, 3),
            expiresAt = Instant.now().minusSeconds(300)
        )
        assertTrue(b.isExpired())
    }

    @Test fun `isExpired returns false for future expiry`() {
        val b = DtnBundle(
            senderUhid = "alice",
            recipientUhid = "bob",
            encryptedPayload = byteArrayOf(1, 2, 3),
            expiresAt = Instant.now().plusSeconds(3600)
        )
        assertFalse(b.isExpired())
    }

    @Test fun `default expiry is in the future`() {
        val b = DtnBundle(
            senderUhid = "alice",
            recipientUhid = "bob",
            encryptedPayload = byteArrayOf(1)
        )
        assertFalse(b.isExpired())
    }

    @Test fun `id is auto-generated UUID when not provided`() {
        val b = DtnBundle(
            senderUhid = "alice",
            recipientUhid = "bob",
            encryptedPayload = byteArrayOf(1)
        )
        assertNotNull(b.id)
    }

    @Test fun `two bundles with different auto-ids are not equal`() {
        val b1 = DtnBundle(senderUhid = "a", recipientUhid = "b", encryptedPayload = byteArrayOf(1))
        val b2 = DtnBundle(senderUhid = "a", recipientUhid = "b", encryptedPayload = byteArrayOf(1))
        assertNotEquals(b1.id, b2.id)
    }
}

// ── BundleStatus enum ─────────────────────────────────────────────────────────

class BundleStatusTest {

    @Test fun `Pending has value 0`() {
        assertEquals(0, BundleStatus.Pending.value)
    }

    @Test fun `InCustody has value 1`() {
        assertEquals(1, BundleStatus.InCustody.value)
    }

    @Test fun `Delivered has value 2`() {
        assertEquals(2, BundleStatus.Delivered.value)
    }

    @Test fun `Expired has value 3`() {
        assertEquals(3, BundleStatus.Expired.value)
    }

    @Test fun `Failed has value 4`() {
        assertEquals(4, BundleStatus.Failed.value)
    }

    @Test fun `fromValue round-trips all statuses`() {
        for (status in BundleStatus.values()) {
            assertEquals(status, BundleStatus.fromValue(status.value))
        }
    }
}

// ── BundlePriority enum ───────────────────────────────────────────────────────

class BundlePriorityTest {

    @Test fun `Low has value 0`() {
        assertEquals(0, BundlePriority.Low.value)
    }

    @Test fun `Normal has value 1`() {
        assertEquals(1, BundlePriority.Normal.value)
    }

    @Test fun `High has value 2`() {
        assertEquals(2, BundlePriority.High.value)
    }

    @Test fun `Sos has value 3`() {
        assertEquals(3, BundlePriority.Sos.value)
    }

    @Test fun `fromValue round-trips all priorities`() {
        for (priority in BundlePriority.values()) {
            assertEquals(priority, BundlePriority.fromValue(priority.value))
        }
    }

    @Test fun `priorities are in increasing order`() {
        assertTrue(BundlePriority.Low.value < BundlePriority.Normal.value)
        assertTrue(BundlePriority.Normal.value < BundlePriority.High.value)
        assertTrue(BundlePriority.High.value < BundlePriority.Sos.value)
    }
}

// ── CustodyRecord ─────────────────────────────────────────────────────────────

class CustodyRecordTest {

    @Test fun `creates with all required fields`() {
        val rec = CustodyRecord(
            bundleId = UUID.randomUUID(),
            fromUhid = "alice",
            toUhid = "bob",
            accepted = true
        )
        assertEquals("alice", rec.fromUhid)
        assertEquals("bob", rec.toUhid)
        assertTrue(rec.accepted)
    }

    @Test fun `auto-generates id when not provided`() {
        val rec = CustodyRecord(
            bundleId = UUID.randomUUID(),
            fromUhid = "a",
            toUhid = "b",
            accepted = false
        )
        assertNotNull(rec.id)
    }

    @Test fun `rejected custody record has accepted=false`() {
        val rec = CustodyRecord(
            bundleId = UUID.randomUUID(),
            fromUhid = "x",
            toUhid = "y",
            accepted = false
        )
        assertFalse(rec.accepted)
    }
}

// ── DtnDeliveryReceipt ────────────────────────────────────────────────────────

class DtnDeliveryReceiptTest {

    @Test fun `creates with all required fields`() {
        val receipt = DtnDeliveryReceipt(
            bundleId = UUID.randomUUID(),
            recipientUhid = "bob",
            totalHops = 4,
            totalCustodyTransfers = 2
        )
        assertEquals("bob", receipt.recipientUhid)
        assertEquals(4, receipt.totalHops)
        assertEquals(2, receipt.totalCustodyTransfers)
    }

    @Test fun `deliveredAt defaults to now`() {
        val before = Instant.now().minusSeconds(1)
        val receipt = DtnDeliveryReceipt(
            bundleId = UUID.randomUUID(),
            recipientUhid = "bob",
            totalHops = 1,
            totalCustodyTransfers = 0
        )
        assertTrue(receipt.deliveredAt >= before)
    }
}

// ── SosAlert ──────────────────────────────────────────────────────────────────

class SosAlertTest {

    @Test fun `creates with senderUhid only`() {
        val alert = SosAlert(senderUhid = "alice")
        assertEquals("alice", alert.senderUhid)
        assertEquals("sos", alert.broadcastType)
        assertNull(alert.message)
    }

    @Test fun `latitude and longitude default to 0`() {
        val alert = SosAlert(senderUhid = "bob")
        assertEquals(0.0, alert.latitude)
        assertEquals(0.0, alert.longitude)
    }

    @Test fun `auto-generates id`() {
        val a1 = SosAlert(senderUhid = "alice")
        val a2 = SosAlert(senderUhid = "alice")
        assertNotEquals(a1.id, a2.id)
    }

    @Test fun `can set message and coordinates`() {
        val alert = SosAlert(
            senderUhid = "alice",
            message = "Need help!",
            latitude = -26.2041,
            longitude = 28.0473,
            broadcastType = "panic"
        )
        assertEquals("Need help!", alert.message)
        assertEquals(-26.2041, alert.latitude)
        assertEquals(28.0473, alert.longitude)
        assertEquals("panic", alert.broadcastType)
    }
}

// ── AetherNode ────────────────────────────────────────────────────────────────

class AetherNodeTest {

    @Test fun `creates with required fields`() {
        val node = AetherNode(
            uhid = "node-001",
            identityPublicKey = ByteArray(32) { it.toByte() }
        )
        assertEquals("node-001", node.uhid)
        assertEquals(32, node.identityPublicKey.size)
    }

    @Test fun `default capabilities are all false`() {
        val node = AetherNode(
            uhid = "n",
            identityPublicKey = ByteArray(32)
        )
        assertEquals(NodeCapabilities(), node.capabilities)
    }

    @Test fun `geohash defaults to null`() {
        val node = AetherNode(uhid = "n", identityPublicKey = ByteArray(32))
        assertNull(node.geohash)
    }

    @Test fun `equals uses content-equality for identityPublicKey`() {
        val key = ByteArray(32) { it.toByte() }
        val n1 = AetherNode("n", key.copyOf())
        val n2 = AetherNode("n", key.copyOf())
        assertEquals(n1, n2)
    }
}

// ── PeerInfo ──────────────────────────────────────────────────────────────────

class PeerInfoTest {

    @Test fun `creates with required fields`() {
        val peer = PeerInfo(
            uhid = "peer-1",
            identityKey = ByteArray(32)
        )
        assertEquals("peer-1", peer.uhid)
        assertEquals(50, peer.reliabilityScore) // default
    }

    @Test fun `geohash defaults to null`() {
        val peer = PeerInfo(uhid = "p", identityKey = ByteArray(32))
        assertNull(peer.geohash)
    }

    @Test fun `equals uses content-equality for identityKey`() {
        val key = ByteArray(32) { 0x42 }
        val p1 = PeerInfo("peer", key.copyOf())
        val p2 = PeerInfo("peer", key.copyOf())
        assertEquals(p1, p2)
    }
}
